Add volume range config

karlvr opened the pull request:
This helps with https://github.com/mopidy/mopidy-alsamixer/issues/3

This doesn't do anything logarithmic with the volume, although it would be easy to do so as the code is pretty simple. As noted in the issue above, perhaps that is best done in pyalsaaudio but making changes there seemed a bit broader! This approach works for me at the moment.

Apologies, this is my first time writing Python.

karlvr commented:
I have now added an option to use a logarithmic volume scale. This seems to work really well. I have kept the option to reduce the effective volume range as well, but perhaps those aren't required anymore and are a bit of a red herring. May still be useful for someone?

adamcik commented on mopidy_alsamixer/__init__.py:
```diff
@@ -22,6 +22,9 @@ def get_config_schema(self):
         schema = super(Extension, self).get_config_schema()
         schema['card'] = config.Integer(minimum=0)
         schema['control'] = config.String()
+        schema['volmin'] = config.Integer(minimum=0, maximum=100)
```
max_volume and min_volume perhaps? Main point being not to abbreviate volume and grouping it together</bikeshed>

adamcik commented on mopidy_alsamixer/__init__.py:
```diff
@@ -22,6 +22,9 @@ def get_config_schema(self):
         schema['control'] = config.String()
+        schema['volmin'] = config.Integer(minimum=0, maximum=100)
+        schema['volmax'] = config.Integer(minimum=0, maximum=100)
+        schema['logarithmic_volume'] = config.Boolean()
```
Should we make this `config.String(choices=('linear', 'logarithmic'))` instead? In case we need to or want to add cubic at some point. Option would be `volume_format` or perhaps `volume_scale`?

karlvr replied:
That's a great suggestion. Honestly I'm a bit unsure as to whether this is logarithmic or cubic :-) I think it is cubic. It seems to work. I'd like to know which it is... I think I'll call it cubic.

karlvr replied:
I've pushed a commit to address this.

adamcik commented on mopidy_alsamixer/mixer.py:
```diff
@@ -70,13 +74,20 @@ def get_volume(self):
-            return int(channels[0])
+            unadjusted_volume = channels[0]
+            if self.logarithmic_volume:
+                unadjusted_volume = 100.0 * math.pow(unadjusted_volume / 100.0, 3.0)
```
FYI: https://lazka.github.io/pgi-docs/#GstAudio-1.0/classes/StreamVolume.html#GstAudio.StreamVolume.convert_volume exists which we could use instead of having the conversion functions our selves. For example:

```
GstAudio.StreamVolume.convert_volume(GstAudio.StreamVolumeFormat.LINEAR, GstAudio.StreamVolumeFormat.DB, 70.0)
```

Also if you don't think we should use the Gst provided one, please make the conversion into an internal function instead of duplicating it again on line 89.

karlvr replied:
Line 89 is actually the inverse of the function on line 79. Line 79 undoes the conversion performed on line 89.

I think using those conversion functions seems like a great idea. I'll make that change and support db (logarithmic) as well. Reading those docs confirms what I implemented was cubic. I mean, there was a `math.pow(x, 3)` in there... so it was pretty obvious.

karlvr replied:
![screen shot 2017-02-03 at 9 06 15 am](https://cloud.githubusercontent.com/assets/1086005/22566308/3498a94e-e9f0-11e6-86f9-0d959f706756.png)

Graph comparing the logarithmic and cubic volume scales. Cubic is in black. Logarithmic in grey.

karlvr replied:
I have pushed a commit using the GstAudio convert_volume function for cubic b12e38bffa6c3727c2c8e1a40ef91e2ea799cac8. That seems to work.

I have also pushed a commit adding logarithmic volume control ad82d9bda7ca6ebc6841c5b1d449aaa819dbad66 but I could not use the convert_volume function as it outputs DB (e.g. negative numbers etc) and we currently use a percentage to control the mixer volume. So I've taken the formula from the GstAudio page (and modified the scaling factor as per my graph above) to give a logarithmic scale, albeit with the formula in our class.

I don't think the logarithmic volume scale is as good as the cubic one! Well, at least it doesn't match up to the alsamixer volume scale like the cubic one does. Still, the cubic one is better!

adamcik commented on mopidy_alsamixer/mixer.py:
```diff
@@ -70,13 +74,20 @@ def get_volume(self):
 
     def set_volume(self, volume):
-        self._mixer.setvolume(volume)
+        adjusted_volume = self.volmin + volume * (self.volmax - self.volmin) / 100.0
```
Turn this into an internal helper function as well?

karlvr replied:
Agreed. Much better. f3fcfed3eb38fe2215410af857c04b514cde58f8

adamcik commented:
https://travis-ci.org/mopidy/mopidy-alsamixer/jobs/197785860 has some errors as the tests don't use the default config (which is kinda silly of us).

https://travis-ci.org/mopidy/mopidy-alsamixer/jobs/197785861 also addresses som style issues that get enforced by our CI.

adamcik commented on README.rst:
```diff
@@ -51,12 +51,23 @@ The following configuration values are available:
+  of ALSA's volume range.
+
+- ``alsamixer/volume_scale``: Either 'linear', 'cubic' or 'db'. The cubic scale is recommended as it produces
+  better results to the ear, and matches the volume scale in alsamixer. The default is the linear scale.
```
It's a bit funny that we say cubic is better, but we use linear anyway... Should we just make cubic the default? Personally I don't use this mixer so I probably shouldn't decide this, but doesn't hurt to ask :-)

karlvr replied:
Let's do it.

adamcik commented on mopidy_alsamixer/mixer.py:
```diff
@@ -70,15 +75,34 @@ def get_volume(self):
+            volume = GstAudio.StreamVolume.convert_volume(GstAudio.StreamVolumeFormat.CUBIC, GstAudio.StreamVolumeFormat.LINEAR, volume / 100.0) * 100.0
+        elif self.volume_scale == "db":
+            volume = math.pow(10, volume / 50.0)
+            #volume = GstAudio.StreamVolume.convert_volume(GstAudio.StreamVolumeFormat.DB, GstAudio.StreamVolumeFormat.LINEAR, volume / 100.0) * 100.0
```
I would just remove "dead" code like this, and maybe add a quick comment about why we calculate this ourselves.

karlvr replied:
Absolutely. Removed.

karlvr commented:
FINALLY! Tests fixed and added to. I added a default config that gets merged with whatever config the tests specify. I've added some tests for cubic volume and log volume.

karlvr commented:
@adamcik just checking in to make sure you saw the latest commits that fixed the tests, and everything else?

adamcik commented:
Thanks for the ping and for adding this, I did indeed miss the previous mails about you getting things fixed.

karlvr commented:
@adamcik thank you, exciting to see this merged. Hope it improves things for people :-)
